subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/DefaultFileSystemAccess.java

```diff
@@ -106,19 +106,17 @@ public <T> Optional<T> readRegularFileContentHash(String location, Function<Hash
                 }
                 HashCode hash = producingSnapshots.guardByKey(location,
                     () -> virtualFileSystem.findSnapshot(location)
-                        .orElseGet(() -> {
+                        .orElseGet(() -> virtualFileSystem.store(location, () -> {
                             HashCode hashCode = hasher.hash(file, fileMetadata.getLength(), fileMetadata.getLastModified());
```

Comment:
Shouldn't we be doing the call to `stat.stat()` in this lambda as well?

Reply:
Yes, it would be better so the whole store action is repeatable. Let's do this in a separate PR though.